File: java-resend-examples/src/main/java/com/resend/examples/Audiences.java
package com.resend.examples;

import com.resend.Resend;
import com.resend.services.audiences.model.ListAudiencesResponseData;
import com.resend.services.contacts.model.CreateContactOptions;
import com.resend.services.contacts.model.CreateContactResponseData;
import com.resend.services.contacts.model.ListContactsResponseData;
import com.resend.services.contacts.model.UpdateContactOptions;
import io.github.cdimascio.dotenv.Dotenv;

public class Audiences {
    public static void main(String[] args) {
        Dotenv dotenv = Dotenv.configure().ignoreIfMissing().load();

        String apiKey = dotenv.get("RESEND_API_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            System.err.println("RESEND_API_KEY environment variable is required");
            System.exit(1);
        }

        Resend resend = new Resend(apiKey);

        String audienceId = dotenv.get("RESEND_AUDIENCE_ID", "your-audience-id");

        try {
            // 1. List audiences
            System.out.println("=== Listing Audiences ===");
            var audiences = resend.audiences().list();
            for (ListAudiencesResponseData audience : audiences.getData()) {
                System.out.println("  - " + audience.getName() + " (" + audience.getId() + ")");
            }

            // 2. Create a contact
            System.out.println("\n=== Creating Contact ===");
            CreateContactOptions createParams = CreateContactOptions.builder()
                    .audienceId(audienceId)
                    .email("clicked@resend.dev")
                    .firstName("Jane")
                    .lastName("Doe")
                    .unsubscribed(false)
                    .build();

            CreateContactResponseData contact = resend.contacts().create(createParams);
            System.out.println("Contact created: " + contact.getId());

            // 3. List contacts
            System.out.println("\n=== Listing Contacts ===");
            var contacts = resend.contacts().list(audienceId);
            for (ListContactsResponseData c : contacts.getData()) {
                System.out.println("  - " + c.getFirstName() + " " + c.getLastName()
                        + " <" + c.getEmail() + "> (unsubscribed: " + c.getUnsubscribed() + ")");
            }

            // 4. Update the contact
            System.out.println("\n=== Updating Contact ===");
            UpdateContactOptions updateParams = UpdateContactOptions.builder()
                    .audienceId(audienceId)
                    .id(contact.getId())
                    .firstName("Janet")
                    .unsubscribed(false)
                    .build();

            resend.contacts().update(updateParams);
            System.out.println("Contact updated: Jane -> Janet");

            // 5. Remove the contact
            System.out.println("\n=== Removing Contact ===");
            resend.contacts().remove(audienceId, contact.getId());
            System.out.println("Contact removed: " + contact.getId());

            System.out.println("\nDone! Full audience/contact lifecycle complete.");
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: elixir-resend-examples/examples/audiences.exs
Mix.install([
  {:resend, "~> 0.4.5"},
  {:dotenvy, "~> 0.8"}
])

if File.exists?(".env"), do: Dotenvy.source!([".env"])

api_key = System.get_env("RESEND_API_KEY") || raise "RESEND_API_KEY environment variable is required"
audience_id = System.get_env("RESEND_AUDIENCE_ID") || "your-audience-id"

# 1. List audiences
IO.puts("=== Listing Audiences ===")

case Resend.Audiences.list() do
  {:ok, %{"data" => audiences}} ->
    Enum.each(audiences, fn audience ->
      IO.puts("  - #{audience["name"]} (#{audience["id"]})")
    end)

  {:error, reason} ->
    IO.puts(:stderr, "Error listing audiences: #{inspect(reason)}")
    System.halt(1)
end

# 2. Create a contact
IO.puts("\n=== Creating Contact ===")

{:ok, contact} =
  Resend.Contacts.create(%{
    audience_id: audience_id,
    email: "clicked@resend.dev",
    first_name: "Jane",
    last_name: "Doe",
    unsubscribed: false
  })

IO.puts("Contact created: #{contact["id"]}")

# 3. List contacts
IO.puts("\n=== Listing Contacts ===")

case Resend.Contacts.list(audience_id) do
  {:ok, %{"data" => contacts}} ->
    Enum.each(contacts, fn c ->
      IO.puts(
        "  - #{c["first_name"]} #{c["last_name"]} <#{c["email"]}> (unsubscribed: #{c["unsubscribed"]})"
      )
    end)

  {:error, reason} ->
    IO.puts(:stderr, "Error listing contacts: #{inspect(reason)}")
end

# 4. Update the contact
IO.puts("\n=== Updating Contact ===")

Resend.Contacts.update(%{
  audience_id: audience_id,
  id: contact["id"],
  first_name: "Janet",
  unsubscribed: false
})

IO.puts("Contact updated: Jane -> Janet")

# 5. Remove the contact
IO.puts("\n=== Removing Contact ===")

Resend.Contacts.remove(audience_id, contact["id"])
IO.puts("Contact removed: #{contact["id"]}")

IO.puts("\nDone! Full audience/contact lifecycle complete.")
